teps/0154-concise-remote-resolver-syntax.md

```diff
@@ -0,0 +1,308 @@
+
+```
+oci://<bundle-provider>/<bundle-namespace>/<bundle-name>@<bundle-version>?kind=<kind>&name=<name>&secret=<secret>
+```
+
+Example
+
+```
+oci://docker.io/ptasci67/example-oci@sha256:053a6cb9f3711d4527dd0d37ac610e8727ec0288a898d5dfbd79b25bcaa29828?kind=task&name=hello-world&secret=default
+```
+
+Using the above example, a user can do the following when referencing the Task:
+```yaml
+taskRef:
+  resolver: bundles
+  name: "oci://docker.io/ptasci67/example-oci@sha256:053a6cb9f3711d4527dd0d37ac610e8727ec0288a898d5dfbd79b25bcaa29828?kind=task&name=hello-world&secret=default"
+```
+
+Occasionally, users may want to fetch from a private registry (e.g. at `10.96.190.208:5000`).
+
+```yaml
+taskRef:
+  resolver: bundles
+  name: "oci://10.96.190.208:5000/ptasci67/example-oci@sha256:053a6cb9f3711d4527dd0d37ac610e8727ec0288a898d5dfbd79b25bcaa29828?kind=task&name=hello-world?secret=default"
+```
+
+##### Http Resolver
+Based on `http` resolver's [params](https://tekton.dev/docs/pipelines/http-resolver/#parameters), we define the following `resolver-reference` format: 
+
+```
+<http-url>?http-username=<http-username>&http-password-secret=<http-password-secret>&http-password-secret-key=<http-password-secret-key>
+```
+
+Here, the url contains the `scheme` and `location`: `<scheme>://<locaiton>`:
+
+Example:
+```
+http://example.com/foo/task.yaml
+https://raw.githubusercontent.com/tektoncd-catalog/git-clone/main/task/git-clone/git-clone.yaml
+```
+
+Using the above example, a user can do the following when referencing the Task:
+```yaml
+taskRef:
+  resolver: http
+  name: "https://raw.githubusercontent.com/tektoncd-catalog/git-clone/main/task/git-clone/git-clone.yaml?http-username=git&http-password-secret=git-secret&http-password-secret-key=git-token"
+```
+
+##### Hub Resolver
+Based on `hub` resolver's [params](https://tekton.dev/docs/pipelines/hub-resolver/#parameters), we define the following `resolver-reference` format: 
+
+```
+hub://<type>/<catalog>/<kind>/<name>@<version>
+```
+**Note** that we use a scheme called `hub` which is only meaningful to the hub resolver.
+
+Example:
+
+```
+hub://artifact/tekton-catalog-tasks/task/git-clone@0.6
+```
+
+Using the above example, a user can do the following when referencing the Task:
+
+```yaml
+taskRef:
+  resolver: hub
+  name: "hub://artifact/tekton-catalog-tasks/task/git-clone@0.6"
+```
+
+##### Cluster Resolver
+Based on `cluster` resolver's [params](https://tekton.dev/docs/pipelines/cluster-resolver/#parameters), we define the following `resolver-reference` format: 
+
+```
+cluster://<kind>/<namespace>/<name>
+```
+
+**Note** that we use a scheme called `cluster` which is only meaningful to the hub resolver.
+
+Example:
+
+```
+cluster://task/default/git-clone
+```
+
+Using the above example, a user can do the following when referencing the Task:
+
+```yaml
+taskRef:
+  resolver: "cluster"
+  name: "cluster://task/default/git-clone" 
+```
+
```

Comment:
Before we get into stage 2, in stage 1 we have multiple resolvers potentially resolving to the same scheme (`http`). What changes (if any) need to be made to the resolution api to support this?

How should a resolver let the controller know that it isn't capable of fulfilling a request?

How is priority determined without phase 2 in place?

Reply:
In phase 1, the `resolver` field is required. So users need to specify the `resolver` and the `name`. This way, the controller knows exactly which resolver to forward the request to. There is no need for priorities or retries. The behaviour is similar to what we have today.

 In phase 2, the `resolver` field  would become **optional** and based soon `name` solely, the controller needs to figure out which resolver to forward the request to. Thats when we need to worry about priority.